lib/spack/spack/solver/asp.py

```diff
@@ -3394,6 +3394,19 @@ def _is_reusable(spec: spack.spec.Spec, packages, local: bool) -> bool:
     return False
 
 
+def _has_runtime_dependencies(spec: spack.spec.Spec) -> bool:
+    if not WITH_RUNTIME:
+        return True
+
+    if spec.satisfies("%gcc") and not spec.satisfies("^gcc-runtime"):
```

Comment:
```suggestion
    if spec.satisfies("%gcc") and not spec.dependencies("gcc-runtime"):
```

Reply:
don't traverse the tree while traversing the tree.

Reply:
(Actually it would be nice to have `has_dependency(...)` which doesn't allocate)

Reply:
Done in https://github.com/spack/spack/pull/43272/commits/b5439b152c9faaa7c87a66c85e72be9de9b33cc8 I didn't do the non-allocating part, but fair to add in another PR.